services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/MultipartManagerImpl.java

```diff
@@ -114,19 +130,49 @@ public ChunkedFileToken createChunkedFileUploadToken(CreateChunkedFileTokenReque
 		cft.setFileName(ccftr.getFileName());
 		cft.setContentType(contentType);
 		cft.setContentMD5(ccftr.getContentMD5());
+		cft.setStorageLocationId(storageLocationId);
 		return cft;
 	}
 	
+	private StorageLocationSetting getStorageLocationSetting(Long storageLocationId) throws DatastoreException, NotFoundException {
+		StorageLocationSetting storageLocationSetting = null;
+		if (storageLocationId != null) {
+			storageLocationSetting = projectSettingsManager.getStorageLocationSetting(storageLocationId);
+		}
+		return storageLocationSetting;
+	}
+
 	/**
 	 * Create a new key
 	 * @param userId
 	 * @param fileName
 	 * @return
 	 */
-	public static String createNewKey(String userId, String fileName) {
-		return String.format(FILE_TOKEN_TEMPLATE, userId, UUID.randomUUID().toString(), fileName);
+	public static String createNewKey(String userId, String fileName, StorageLocationSetting storageLocationSetting) {
+		String base = "";
+		if (storageLocationSetting instanceof ExternalS3StorageLocationSetting) {
+			ExternalS3StorageLocationSetting externalS3StorageLocationSetting = (ExternalS3StorageLocationSetting) storageLocationSetting;
+			if (!StringUtils.isEmpty(externalS3StorageLocationSetting.getBaseKey())) {
+				base = externalS3StorageLocationSetting.getBaseKey() + FILE_TOKEN_TEMPLATE_SEPARATOR;
+			}
+		}
+		return String.format(FILE_TOKEN_TEMPLATE, base, userId, UUID.randomUUID().toString(), fileName);
 	}
 	
+	private static String getBucket(StorageLocationSetting storageLocationSetting) {
+		String bucket = StackConfiguration.getS3Bucket();
+		if (storageLocationSetting instanceof ExternalS3StorageLocationSetting) {
+			bucket = ((ExternalS3StorageLocationSetting) storageLocationSetting).getBucket();
+		}

```

Comment:
need to handle the case in which the argument is NEITHER ExternalS3StorageLocationSetting NOR S3StorageLocationSetting. In its current form anything that's not ExternalS3StorageLocationSetting is assumed to be S3StorageLocationSetting.